apiserver/client/status.go

```diff
@@ -493,21 +497,23 @@ func fetchAllApplicationsAndUnits(
 	if err != nil {
 		return nil, nil, nil, err
 	}
-	for _, s := range applications {
-		units, err := s.AllUnits()
-		if err != nil {
-			return nil, nil, nil, err
-		}
+	units, err := model.AllUnits()
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	for _, app := range applications {
 		appUnitMap := make(map[string]*state.Unit)
 		for _, u := range units {
-			appUnitMap[u.Name()] = u
+			if u.ApplicationName() == app.Name() {
```

Comment:
3000 units in a model with 190 applications becomes 190*3000 to evaluate this loop.
If you want to do this, we need to first put the units into a map on app name => units. (which happens to be what I was specifically testing, ~190 apps, each with 15 units.)
That may not be the specific layout, but given you can convert this from O(N*M) to O(N*1) by putting it into a hash table, I think its better to do so.